src/Combinatorics/EnumerativeCombinatorics/partitions.jl

```diff
@@ -332,121 +364,356 @@ julia> collect(partitions(7, 3, 1, 4; only_distinct_parts = true))
   end
 
-  # If l1 == 0 the algorithm parta will actually create lists containing the
-  # entry zero, e.g. partitions(2, 2, 0, 2) will contain [2, 0].
-  # This is nonsense, so we set l1 = 1 in this case.
-  if l1 == 0
-    l1 = 1
+  if k == 1 && lb <= n <= ub
+    return partition(T[n], check=false), (T[], T[], 0, 0, 1, false)
   end
 
-  #Algorithm starts here
-  P = Partition{T}[]    #this will be the array of all partitions
-  x = zeros(T, n)
-  y = zeros(T, n)
-  j = only_distinct_parts*n*(n - 1)
-  m = m - n*l1 - div(j, 2)
-  l2 = l2 - l1
-  if 0 <= m <= n*l2 - j
+  x = zeros(T,k)
+  y = zeros(T,k)
+  jj = only_distinct_parts*k*(k-1)
+  N = n - k*lb - div(jj,2)
+  L2 = ub-lb
+  0 <= N <= k*L2 - jj || return nothing
 
-    for i = 1:n
-      y[i] = x[i] = l1 + only_distinct_parts*(n - i)
-    end
+  for i in 1:k
+    y[i] = x[i] = lb + only_distinct_parts*(k-i)
+  end
 
-    i = 1
-    l2 = l2 - only_distinct_parts*(n - 1)
+  i = 1
+  L2 = L2 - only_distinct_parts*(k-1)
 
-    while true
-      while m > l2
-        m -= l2
-        x[i] = y[i] + l2
-        i += 1
-      end
+  while N > L2
+    N -= L2
+    x[i] = y[i] + L2
+    i += 1
+  end
+  x[i] = y[i] + N
+  return partition(x[1:k], check = false), (x, y, N, L2, i, true)
+end
 
-      x[i] = y[i] + m
-      push!(P, partition(x[1:n], check = false))
+function Base.iterate(P::PartitionsFixedNumParts{T}, state::Tuple{Vector{T}, Vector{T}, T, IntegerUnion, Int, Bool}) where T
+  k = P.k
+  x, y, N, L2, i, flag = state
 
-      if i < n && m > 1
-        m = 1
-        x[i] = x[i] - 1
-        i += 1
-        x[i] = y[i] + 1
-        push!(P, partition(x[1:n], check = false))
-      end
+  N == 0 && return nothing
 
-      lcycle = false
-      for j = i - 1:-1:1
-        l2 = x[j] - y[j] - 1
-        m = m + 1
-        if m <= (n - j)*l2
-          x[j] = y[j] + l2
-          lcycle = true
-          break
-        end
-        m = m + l2
-        x[i] = y[i]
-        i = j
-      end
+  if flag
+    if i < k && N > 1
+      N = 1
+      x[i] = x[i] - 1
+      i += 1
+      x[i] = y[i] + 1
+      return partition(x[1:k], check = false), (x,y,N,L2,i,false)
+    end
+  end
 
-      if !lcycle
-        break
-      end
+  lcycle = false
+  for j in i - 1:-1:1
+    L2 = x[j] - y[j] - 1
+    N = N + 1
+    if N <= (k-j)*L2
+      x[j] = y[j] + L2
+      lcycle = true
+      break
     end
+    N = N + L2
+    x[i] = y[i]
+    i = j
   end
 
-  return (p for p in P)
+  lcycle || return nothing
+  while N > L2
+    N -= L2
+    x[i] = y[i] + L2
+    i += 1
+  end
+  x[i] = y[i] + N
+  return partition(x[1:k], check = false), (x,y,N,L2,i,true)
 end
 
-function partitions(m::T, n::IntegerUnion; only_distinct_parts::Bool = false) where T <: IntegerUnion
-  @req m >= 0 "m >= 0 required"
+
+# function partitions(n::T, k::IntegerUnion, lb::IntegerUnion, ub::IntegerUnion; only_distinct_parts::Bool = false) where T <: IntegerUnion
```

Comment:
Any convincing reason to keep this commented out code? Otherwise just delete it...?

Reply:
I migrated some of the comments, and deleted the code.